flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionController.java

```diff
@@ -0,0 +1,241 @@
+     * @param state the state to request.
+     * @param type the type of this request.
+     * @param payload the payload input for this request.
+     * @return the state future.
+     */
+    public <IN, OUT> InternalStateFuture<OUT> handleRequest(
+            @Nullable State state, StateRequest.RequestType type, @Nullable IN payload) {
+        // Step 1: build state future & assign context.
+        InternalStateFuture<OUT> stateFuture = stateFutureFactory.build(currentContext);
+
+        StateRequest<K, IN, OUT> request =
+                new StateRequest<>(state, type, payload, stateFuture, currentContext);
+
+        // Step 2: try to seize the capacity, if the current in-flight records exceeds the limit,
+        // block the current state request from entering until some buffered requests are processed.
+        seizeCapacity();
+
+        // Step 3: try to occupy the key and place it into right buffer.
+        if (currentContext.tryOccupyKey()) {
+            insertActiveBuffer(request);
+        } else {
+            insertBlockingBuffer(request);
+        }
+        // Step 4: trigger the (active) buffer if needed.
+        triggerIfNeeded(false);
+        return stateFuture;
+    }
+
+    <IN, OUT> void insertActiveBuffer(StateRequest<K, IN, OUT> request) {
+        activeBuffer.add(request);
+    }
+
+    <IN, OUT> void insertBlockingBuffer(StateRequest<K, IN, OUT> request) {
+        blockingBuffer.add(request);
+    }
+
+    /**
+     * Trigger a batch of requests.
+     *
+     * @param force whether to trigger requests in force.
+     */
+    @VisibleForTesting
+    void triggerIfNeeded(boolean force) {
+        if (!force && activeBuffer.size() < batchSize) {
+            return;
+        }
+        migrateBlockingToActive();
+        List<StateRequest<?, ?, ?>> toRun = new LinkedList<>(activeBuffer);
+        activeBuffer.clear();
+        stateExecutor.executeBatchRequests(toRun);
+        // migrate state requests in blocking buffer to active buffer actively, let the requests
+        // that enter the buffer first be processed first.
+        migrateBlockingToActive();
+    }
+
+    /** Migrate the blocking requests to the active buffer. */
+    @VisibleForTesting
+    void migrateBlockingToActive() {
+        Iterator<StateRequest<K, ?, ?>> blockingIter = blockingBuffer.iterator();
+        while (blockingIter.hasNext()) {
+            StateRequest<K, ?, ?> stateRequest = blockingIter.next();
+            if (stateRequest.getRecordContext().tryOccupyKey()) {
+                insertActiveBuffer(stateRequest);
+                blockingIter.remove();
+            }
+        }
+    }
+
+    private void seizeCapacity() {
+        // 1. Check if the record is already in buffer. If yes, this indicates that it is a state
+        // request resulting from a callback statement, otherwise, it signifies the initial state
+        // request
+        // for a newly entered record.
+        if (currentContext.isInBuffer()) {
```

Comment:
Is `keyOccupied` enough for this senario?

Reply:
`keyOccupied` is enough, thanks for the suggestion.